build-internals/build.zig

```diff
@@ -0,0 +1,265 @@
+            .imports = decl.imports,
+        });
+    }
+};
+
+/// The resulting binary format for the firmware file.
+/// A lot of embedded systems don't use plain ELF files, thus we provide means
+/// to convert the resulting ELF into other common formats.
+pub const BinaryFormat = union(enum) {
+    /// [Executable and Linkable Format](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format), the standard output from the compiler.
+    elf,
+
+    /// A flat binary, contains only the loaded portions of the firmware with an unspecified base offset.
+    bin,
+
+    /// The [Intel HEX](https://en.wikipedia.org/wiki/Intel_HEX) format, contains
+    /// an ASCII description of what memory to load where.
+    hex,
+
+    /// A [Device Firmware Upgrade](https://www.usb.org/sites/default/files/DFU_1.1.pdf) file.
+    dfu,
+
+    /// The [USB Flashing Format (UF2)](https://github.com/microsoft/uf2) designed by Microsoft.
+    uf2: enum {
+        ATMEGA32,
+        SAML21,
+        NRF52,
+        ESP32,
+        STM32L1,
+        STM32L0,
+        STM32WL,
+        LPC55,
+        STM32G0,
+        GD32F350,
+        STM32L5,
+        STM32G4,
+        MIMXRT10XX,
+        STM32F7,
+        SAMD51,
+        STM32F4,
+        FX2,
+        STM32F2,
+        STM32F1,
+        NRF52833,
+        STM32F0,
+        SAMD21,
+        STM32F3,
+        STM32F407,
+        STM32H7,
+        STM32WB,
+        ESP8266,
+        KL32L2,
+        STM32F407VG,
+        NRF52840,
+        ESP32S2,
+        ESP32S3,
+        ESP32C3,
+        ESP32C2,
+        ESP32H2,
+        RP2040,
+        RP2XXX_ABSOLUTE,
+        RP2XXX_DATA,
+        RP2350_ARM_S,
+        RP2350_RISC_V,
+        RP2350_ARM_NS,
+        STM32L4,
+        GD32VF103,
+        CSK4,
+        CSK6,
+        M0SENSE,
+    },
+
+    /// The [firmware format](https://docs.espressif.com/projects/esptool/en/latest/esp32/advanced-topics/firmware-image-format.html) used by the [esptool](https://github.com/espressif/esptool) bootloader.
+    esp,
+
+    /// Custom option for non-standard formats.
+    custom: *Custom,
+
+    /// Returns the standard extension for the resulting binary file.
+    pub fn get_extension(format: BinaryFormat) []const u8 {
+        return switch (format) {
+            .elf => ".elf",
+            .bin => ".bin",
+            .hex => ".hex",
+            .dfu => ".dfu",
+            .uf2 => ".uf2",
+            .esp => ".bin",
+
+            .custom => |c| c.extension,
+        };
+    }
+
+    pub const Custom = struct {
+        /// The standard extension of the format.
+        extension: []const u8,
+
+        /// A function that will convert a given `elf` file into the custom output format.
+        ///
+        /// The `*Custom` format is passed so contextual information can be obtained by using
+        /// `@fieldParentPtr` to provide access to tooling.
+        convert: *const fn (*Custom, elf: Build.LazyPath) Build.LazyPath,
+    };
```

Comment:
`Custom` needs the providing `*std.Dependency` or `*std.Build` otherwise we can't allocate memory or access custom tooling (via `b.dependency` or `b.artifact`)